File: image-crud/src/main/java/org/judexmars/imagecrud/config/healthcheck/KafkaHealthIndicator.java
package org.judexmars.imagecrud.config.healthcheck;

import java.util.concurrent.ExecutionException;
import lombok.RequiredArgsConstructor;
import org.apache.kafka.clients.admin.AdminClient;
import org.springframework.boot.actuate.health.Health;
import org.springframework.boot.actuate.health.HealthIndicator;
import org.springframework.kafka.core.KafkaAdmin;
import org.springframework.stereotype.Component;

/**
 * Health indicator for kafka brokers.
 */
@Component
@RequiredArgsConstructor
public class KafkaHealthIndicator implements HealthIndicator {

  private final KafkaAdmin kafkaAdmin;

  @Override
  public Health health() {
    try (var adminClient = AdminClient.create(kafkaAdmin.getConfigurationProperties())) {
      var describeClusterResult = adminClient.describeCluster();
      int nodeCount = describeClusterResult.nodes().get().size();
      if (nodeCount > 0) {
        return Health.up().withDetail("nodeCount", nodeCount).build();
      } else {
        return Health.down().withDetail("nodeCount", nodeCount).build();
      }
    } catch (InterruptedException | ExecutionException e) {
      return Health.down(e).build();
    }
  }
}

File: image-processor/src/main/kotlin/org/judexmars/imageprocessor/config/healthcheck/KafkaHealthIndicator.kt
package org.judexmars.imageprocessor.config.healthcheck

import org.apache.kafka.clients.admin.AdminClient
import org.springframework.boot.actuate.health.Health
import org.springframework.boot.actuate.health.HealthIndicator
import org.springframework.kafka.core.KafkaAdmin
import org.springframework.stereotype.Component

/**
 * Health indicator for kafka brokers.
 */
@Component
class KafkaHealthIndicator(
    private val kafkaAdmin: KafkaAdmin,
) : HealthIndicator {
    override fun health(): Health {
        try {
            AdminClient.create(kafkaAdmin.configurationProperties).use { adminClient ->
                val describeClusterResult = adminClient.describeCluster()
                val nodeCount = describeClusterResult.nodes().get().size
                return if (nodeCount > 0) {
                    Health.up().withDetail("nodeCount", nodeCount).build()
                } else {
                    Health.down().withDetail("nodeCount", nodeCount).build()
                }
            }
        } catch (e: Exception) {
            return Health.down(e).build()
        }
    }
}
